### PR TITLE
Remove assignments with strengths weaker than strongest non-tristate RHS 

### PR DESCRIPTION
**Pull request by RRozak**
Adds another step of static resolution of signal strengths.
After tristate propagation some statements may be known to not equal `z`. If such a statement exists, all weaker statements can be safely removed. Since the value of the RHS is generally unknown (the known values with all bits being the same), weaker of the strengths has to be taken into account.
So this step consists of just finding the strongest assignment with non-tristate RHS and removing all weaker or equal to it.

Assignments with RHS marked as tristates can't be handled in such way, because if their RHS are equal to `z`, they should be skipped and it is generally not known at verilation time.

I also added more tests to test the step added previously: https://github.com/antmicro/verilator-1/blob/610d30f6cc46f619fb4a8b7a8ef7f15b4591bad3/test_regress/t/t_strength_strongest_constant.v.
This test:
```systemverilog
  wire [1:0] b;
   assign (weak0, supply1) b = '1;
   assign b = clk1 ? '0 : 'z;
```
throws error currently on master, because vertices of tristate graph aren't removed from the whole subtree of removed assignment, but only from its RHS node.

### CODE REVIEW

File: src/V3Tristate.cpp [hunk 1]
```diff
@@ -269,6 +277,17 @@ class TristateGraph final {
...
+    void deleteVerticesFromSubtree(AstNode* nodep) {
+        // Skip vars, because they may be connected to more than one varref
+        if (!VN_IS(nodep, Var)) {
+            TristateVertex* const vertexp = reinterpret_cast<TristateVertex*>(nodep->user5p());
+            if (vertexp) vertexp->unlinkDelete(&m_graph);
+        }
+        if (nodep->op1p()) deleteVerticesFromSubtree(nodep->op1p());
+        if (nodep->op2p()) deleteVerticesFromSubtree(nodep->op2p());
+        if (nodep->op3p()) deleteVerticesFromSubtree(nodep->op3p());
+        if (nodep->op4p()) deleteVerticesFromSubtree(nodep->op4p());
```
Comment (kbieganski) [inline src/V3Tristate.cpp:289]: Might be better to put `if (!nodep) return;` at the start and remove the `if`s?

File: src/V3Tristate.cpp [hunk 2]
```diff
@@ -715,64 +734,92 @@ class TristateVisitor final : public TristateBaseVisitor {
...
+                m_tgraph.deleteVerticesFromSubtree(assignp);
+                VL_DO_DANGLING(pushDeletep(assignp->unlinkFrBack()), assignp);
+                return true;
+            }
+            return false;
+        });
+        assigns.erase(removedIt, assigns.end());
+    }
+
+    void removeAssignmentsNotStrongerThanUniformConstant() {
```
Comment (kbieganski) [inline src/V3Tristate.cpp:768]: I'm afraid some of these names are a bit too long :) What is the uniform constant?

**Review by kbieganski — COMMENTED** [review]
Looks OK to me, just some comments. I also edited the description a bit.

File: src/V3Tristate.cpp (hunk 2 above)
Comment (RRozak) [reply src/V3Tristate.cpp:768]: I called it a constant equal to '0 or to '1.
I know that they are long, but I wanted them to be more self-descriptive and I couldn't find anything shorter. But if you have any ideas, please tell me.

**Pull request closed without merging**